File: palindrome_permutation.py
"""Given a string, write a function to determine if it is a permutation of a palindrome
"""
# A permutation of a palindrome would be something where:
# - If there is an even number of letters, 
#   each letter should appear an even number of times
# - If there is an odd number on letters,
#   each letter *except one* should appear an even amount of times

def preprocess(string):
    string = "".join(string.split()).lower()
    pro_string = "".join(sorted(list(string)))
    return str(pro_string)

def is_palindrome(string):
    s = preprocess(string)
    if (s.__len__() % 2) == 0:
        for i in range(0, s.__len__()-1, 2):
            if (s[i] != s[i+1]):
                return False
        return True
    else:
        # If there is an odd number...
        unpaireds_found = 0
        i = 0
        for c in range(0, s.__len__()-1, 2):
            if (s[i] != s[i+1] and unpaireds_found > 0):
                return False
            elif(s[i] != s[i+1]):
                unpaireds_found += 1
                i += 1
            else:
                i += 2
        return True


"""TEST METHODS"""

def returns_true_when_pal_odd():
    if is_palindrome("atco cta") == True:
        return 1
    else:
        return 0

def returns_true_when_pal_even():
    if is_palindrome("car cra") == True:
        return 1
    else:
        return 0

def returns_false_when_not_pal_odd():
    if is_palindrome("arbor bear") == False:
        return 1
    else:
        return 0

def returns_false_when_not_pal_even():
    if is_palindrome("dog bug") == False:
        return 1
    else:
        return 0

def preprocess_works_correctly():
    if preprocess("  a rbO R beA     r  ") == "aabbeorrr":
        return 1
    else:
        return 0

def run_tests():
    print(returns_true_when_pal_odd())
    print(returns_true_when_pal_even())
    print(returns_false_when_not_pal_odd())
    print(returns_false_when_not_pal_even())
    print(preprocess_works_correctly())

if __name__ == "__main__":
    run_tests()

File: PalindromePermJava/palindromePermutation.java
/**
 * Given a string, write a function to determine if it is a permutation of a palindrome
 */

import java.util.*;

public class palindromePermutation
{
    public static void main(String[] args) {
        System.out.println(isPalRetsTrueOdd());
        System.out.println(isPalRetsTrueEven());
        System.out.println(isPalRetsFalseOdd());
        System.out.println(isPalRetsFalseEven());
        System.out.println(preprocessWorks());
    }

    // A permutation of a palindrome would be something where:
    // - If there is an even number of letters, 
    //   each letter should appear an even number of times
    // - If there is an odd number on letters,
    //   each letter *except one* should appear an even amount of times

    public static boolean isPal(String str) {
        char[] proStr = preprocess(str);
        if(proStr.length % 2 == 0) {
            // If even number of characters:
            for(int i = 0; i < proStr.length-1; i+=2) {
                if(proStr[i] != proStr[i+1]) {
                    return false;
                }
            }
            return true;
        } else {
            int unpairedsFound = 0;
            for(int i = 0; i < proStr.length-1; i+=2) {
                if((proStr[i] != proStr[i+1]) && (unpairedsFound > 0)) {
                    return false;
                } else if(proStr[i] != proStr[i+1]) {
                    unpairedsFound++;
                    i--;
                }
            }
            return true;
        }
    }

    public static char[] preprocess(String str) {
        char[] proStr = str.toLowerCase().replaceAll("\\s+","").toCharArray();
        Arrays.sort(proStr);
        return proStr;
    }

    /* TEST METHODS */

    public static int isPalRetsTrueOdd() {
        if(isPal("a cto atc") == true) {
            return 1;
        }
        return 0;
    }

    public static int isPalRetsTrueEven() {
        if(isPal("rat tra") == true) {
            return 1;
        }
        return 0;
    }

    public static int isPalRetsFalseOdd() {
        if(isPal("arbor bear") == false) {
            return 1;
        }
        return 0;
    }

    public static int isPalRetsFalseEven() {
        if(isPal("mean door") == false) {
            return 1;
        }
        return 0;
    }

    public static int preprocessWorks() {
        if(Arrays.equals(preprocess("   a r B  OrB E a r "), "aabbeorrr".toCharArray())) {
            return 1;
        }
        return 0;
    }
}